tests/integration/security/sds_ingress/util/util.go

```diff
@@ -0,0 +1,256 @@
+		},
+	}
+}
+
+type ExpectedResponse struct {
+	ResponseCode int
+	ErrorMessage string
+}
+
+// VisitProductPage makes HTTPS request to ingress gateway to visit product page
+func VisitProductPage(ingress ingress.Instance, host string, timeout time.Duration, exRsp ExpectedResponse, t *testing.T) error {
+	start := time.Now()
+	for {
+		response, err := ingress.Call("/productpage", host)
+		errorMatch := true
+		if err != nil {
+			t.Logf("Unable to connect to product page: %v", err)
+			if !strings.Contains(err.Error(), exRsp.ErrorMessage) {
+				errorMatch = false
+			}
+		}
+
+		status := response.Code
+		if status == exRsp.ResponseCode && errorMatch {
+			t.Logf("Got %d response from product page!", status)
+			return nil
+		} else if status != exRsp.ResponseCode {
+			t.Errorf("expected response code %d but got %d", exRsp.ResponseCode, status)
+		} else {
+			t.Errorf("expected response error message %s but got %s", exRsp.ErrorMessage, err.Error())
+		}
+
+		if time.Since(start) > timeout {
+			return fmt.Errorf("could not retrieve product page in %v: Last status: %v", timeout, status)
+		}
+
+		time.Sleep(3 * time.Second)
+	}
+}
+
+// RotateSecrets deletes kubernetes secrets by name in credNames and creates same secrets using key/cert
+// from ingressCred.
+func RotateSecrets(t *testing.T, ctx framework.TestContext, credNames []string,
+	ingressType ingress.GatewayType, ingressCred IngressCredential) {
+	istioCfg := istio.DefaultConfigOrFail(t, ctx)
+	systemNS := namespace.ClaimOrFail(t, ctx, istioCfg.SystemNamespace)
+	kubeAccessor := ctx.Environment().(*kube.Environment).Accessor
+	for _, cn := range credNames {
+		err := kubeAccessor.DeleteSecret(systemNS.Name(), cn)
+		if err != nil {
+			t.Errorf("Failed to delete secret (error: %s)", err)
+		}
+	}
+	// Check if Kubernetes secret is deleted
+	maxRetryNumber := 5
+	checkRetryInterval := time.Second * 1
+	for _, cn := range credNames {
+		t.Logf("Check ingress Kubernetes secret %s:%s...", systemNS.Name(), cn)
+		for i := 0; i < maxRetryNumber; i++ {
+			_, err := kubeAccessor.GetSecret(systemNS.Name()).Get(cn, metav1.GetOptions{})
+			if err != nil {
+				t.Logf("Secret %s:%s is deleted", systemNS.Name(), cn)
+				break
+			} else {
+				t.Logf("Secret %s:%s still exists.", systemNS.Name(), cn)
+				time.Sleep(checkRetryInterval)
+			}
+		}
+	}
+
+	CreateIngressKubeSecret(t, ctx, credNames, ingressType, ingressCred)
+}
+
+// DeployBookinfo deploys bookinfo application, and deploys gateway with various type.
+// nolint: interfacer
+func DeployBookinfo(t *testing.T, ctx framework.TestContext, g galley.Instance, gatewayType GatewayType) {
+	bookinfoNs, err := namespace.New(ctx, "istio-bookinfo", true)
+	if err != nil {
+		t.Fatalf("Could not create istio-bookinfo Namespace; err:%v", err)
+	}
+	d := bookinfo.DeployOrFail(t, ctx, bookinfo.Config{Namespace: bookinfoNs, Cfg: bookinfo.BookInfo})
+
+	env.BookInfoRoot = path.Join(env.IstioRoot, "tests/integration/security/sds_ingress/")
```

Comment:
This doesn't look right. You're changing the value of a global static here.

Reply:
This setup follows the task guide here https://istio.io/docs/tasks/traffic-management/secure-ingress/mount/#configure-traffic-for-the-bookinfo-com-host

It first deploys bookinfo service, and then change the bookinforoot in order to deploy gateway and virtual service. 
I add a originBookInfoRoot to save the original env.BookInfoRoot, and restore the env.BookInfoRoot at the end of the method.